src/config/secret.rs

```diff
@@ -0,0 +1,241 @@
+    re.captures_iter(input).for_each(|cap| {
+        if let (Some(backend), Some(key)) = (cap.get(1), cap.get(2)) {
+            if let Some(keys) = keys.get_mut(backend.as_str()) {
+                keys.push(key.as_str().to_string());
+            } else {
+                keys.insert(backend.as_str().to_string(), vec![key.as_str().to_string()]);
+            }
+        }
+    });
+    keys
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ExecBackend {
+    pub command: Vec<String>,
+    #[serde(default = "default_timeout_secs")]
+    pub timeout: u64,
+}
+
+const fn default_timeout_secs() -> u64 {
+    5
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ExecQuery {
+    version: String,
+    secrets: Vec<String>,
+}
+
+fn new_query(secrets: Vec<String>) -> ExecQuery {
+    ExecQuery {
+        version: "1.0".to_string(),
+        secrets,
+    }
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone)]
+struct ExecResponse {
+    value: Option<String>,
+    error: Option<String>,
+}
+
+#[typetag::serde(name = "exec")]
+impl SecretBackend for ExecBackend {
+    fn retrieve(&mut self, secret_keys: Vec<String>) -> crate::Result<HashMap<String, String>> {
+        let mut output = executor::block_on(tokio::time::timeout(
+            Duration::from_secs(self.timeout),
+            query_backend(&self.command, new_query(secret_keys.clone())),
+        ))??;
+        let mut secrets = HashMap::new();
+        for k in secret_keys.into_iter() {
+            if let Some(secret) = output.get_mut(&k) {
+                if let Some(e) = &secret.error {
+                    return Err(format!("secret for key '{}' was not retrieved: {}", k, e).into());
+                }
+                if let Some(v) = secret.value.take() {
+                    if v.is_empty() {
+                        return Err(format!("secret for key '{}' was empty", k).into());
+                    }
+                    secrets.insert(k, v);
+                } else {
+                    return Err(format!("secret for key '{}' was empty", k).into());
+                }
+            } else {
+                return Err(format!("secret for key '{}' was not retrieved", k).into());
+            }
+        }
+        Ok(secrets)
+    }
+}
+
+async fn query_backend(
+    cmd: &[String],
+    query: ExecQuery,
+) -> crate::Result<HashMap<String, ExecResponse>> {
+    let command = &cmd[0];
+    let mut command = Command::new(command);
+
+    if cmd.len() > 1 {
+        command.args(&cmd[1..]);
+    };
+
+    command.kill_on_drop(true);
+    command.stderr(std::process::Stdio::piped());
+    command.stdin(std::process::Stdio::piped());
+    command.stdout(std::process::Stdio::piped());
+
+    let mut child = command.spawn()?;
+    let mut stdin = child.stdin.take().ok_or("unable to acquire stdin")?;
+
+    let query = serde_json::to_vec(&query)?;
+    stdin.write_all(&query).await?;
+    drop(stdin);
+
+    let output = child.wait_with_output().await?;
```

Comment:
We should probably wire up a shutdown signal here so that if a user ctrl-C's, it aborts rather than waiting for the secret command to finish. I imagine the topology shutdown signal doesn't exist yet, but maybe another already does?